Developers guide (#40)
websight-io/docs: docs/getting-started/dev-quick-start/index.md

# Developers Quick Start Guide

## Introduction

Use this guide to learn how to update a simple Howlite component. We recommend completing the general [quick start guide](/docs/getting-started/quick-start/index.md) first to get familiar with basic concepts.

> The scenario utilises a sample component library [Howlite](https://github.com/websight-io/howlite) and a demo site project _Luna_. For simplicity, we overrode the _Title_ component included in the [Howlite](https://github.com/websight-io/howlite) collection in advance and prepared the _Luna Title_ component inside the demo project. 

Reviewer: ```suggestion
The scenario utilizes a sample component library [Howlite](https://github.com/websight-io/howlite) and a demo site project _Luna_. For simplicity, we overrode the _Title_ component included in the [Howlite](https://github.com/websight-io/howlite) collection in advance and prepared the _Luna Title_ component inside the demo project. 
```

Your task is to update the _Luna Title_ component and extend existing functionality and tests. See more details in the sections below.

!!! info "Important notice"

    All scripts presented below are designed for Linux-based platforms. If you are a Windows user, please install and configure [Windows Subsystem for Linux](https://learn.microsoft.com/en-us/windows/wsl/install) (WSL).

## Part A: Prerequisites

1. Install [AdoptOpenJDK 17](https://adoptium.net/) with 'x64/aarch64' architecture (on mac use `brew install openjdk@17`):
``` script
$ java --version
openjdk 17.0.2 2022-01-18
OpenJDK Runtime Environment Temurin-17.0.2+8 (build 17.0.2+8)
OpenJDK 64-Bit Server VM Temurin-17.0.2+8 (build 17.0.2+8, mixed mode)
```
1. Install Maven 3.8.5+

1. Install [Docker Desktop](https://www.docker.com/products/docker-desktop/).

## Part B: Setup local environment

> ### Download Maven configuration

Reviewer 2: This step can be avoided if we define the `repository` in the `luna-parent` pom.xml:

```
  <repositories>
    <repository>
      <id>websight-repo</id>
      <url>https://repo.websight.io/maven/</url>
      <snapshots>
        <enabled>false</enabled>
      </snapshots>
    </repository>
  </repositories>
```

Reviewer 3: @malaskowski great! let's do it. please update me when ready and I will remove this part.

Reviewer 3: done

Create a dedicated settings file to not override your default configuration.

``` shell
mkdir -p ~/.m2
curl -s https://repo.websight.io/settings-websight-public.xml --output ~/.m2/websight-settings.xml
```

### Build Luna

Clone _Luna_ repository and build it.

```shell
git clone https://github.com/websight-io/luna-project.git
cd luna-project

> mvn clean install -s ~/.m2/websight-settings.xml -P e2e

Author: ```suggestion
mvn clean install -P e2e
```

```

### Run docker environment
```shell
docker compose -f environment/docker-compose.yml up -d
```

> Congratulations! Your local environment is ready. Open [http://localhost:8080/](http://localhost:8080/) and login using `wsadmin`/`wsadmin` credentials.

Reviewer: ```suggestion
Congratulations! Your local environment is ready. Open [http://localhost:8080/](http://localhost:8080/) and log in using `wsadmin`/`wsadmin` credentials.
```

## Part C: Changing component

!!! info "Hint"

>     If you need help to navigate inside WebSight see the general [getting started](/docs/getting-started/quick-start/index.md) for details.

Reviewer: ```suggestion
    If you need help to navigate inside WebSight, see the general [getting started](/docs/getting-started/quick-start/index.md) for details.
```

### Business requirement

Let's imagine the following scenario. Page owner wants to update the title of the new Grand Luxor Jewelry Collection.

![The current title](the-current-title.png)

The ask is to keep the collection name in one line and decrease the font size for the text `Meet our`. The expected result is presented below.

![Expected result](the-expected-result.png)

### Technical scope

You need to extend the _Title_ component included in the [Howlite](https://github.com/websight-io/howlite) to implement the above requirement. However, let's check the orginal component first to identify the scope of changes. 

> Run WebSight, open _Luna_ space and edit home page. Find the _Title_ having text `Meet our New Grand Luxor Jewelry Collection` and edit its properties. 

Reviewer: ```suggestion
Run WebSight, open _Luna_ space and edit home page. Find the _Title_ with text `Meet our New Grand Luxor Jewelry Collection` and edit its properties. 
```

![Title dialog properties](title-dialog-properties.png)

> Enable `Overline text` option, move `Meet our` from `Heding text` to `Overline text` and submit changes.

Reviewer: ```suggestion
Enable `Overline text` option, move `Meet our` from `Heading text` to `Overline text` and submit changes.
```

```json title="luna/core/src/main/resources/apps/luna/components/lunatitle/dialog/.content.json"
{
  "tabs": {
    "generalTab": {
      "container": {
        "overlineSize": {
          "sling:resourceType": "wcm/dialogs/components/include",
          "sling:orderBefore": "overline",
          "path": "/libs/howlite/components/common/headingsize",
          "include": {
            "sling:resourceSuperType": "/libs/howlite/components/common/headingsize",
            "label": "Overline size",
            "name": "overlineSize",
            "description": "Changes font size",
            "s": {
              "selected": true
            },
            "m": {
              "selected": false
            }
          }
        }
      }
    }
  }
}
```

The above definition specifies the new `overlineSize` field. It is placed before `overline` field and uses heading size definition from Howlite, but with _small_ size selected by default.

### Install changes 
Run the following command to install the changes on your local environment.
```shell
mvn -f luna/core/pom.xml clean install -P autoInstallBundle
```

## Part D: Functional tests

> ### Running functional tests

Reviewer: ```suggestion
### Run functional tests
```

```

Run them using the following command.
```shell
npm run-script test --prefix tests/end-to-end
```

If you execute them, they detect your changes for the _Luna Title_ and fail. You should get the following results.
```
Running:  lunatitle.cy.ts                                                                 (1 of 1)

Luna Title component
  1) renders correctly in preview mode
  2) renders correctly in edit mode

0 passing (8s)
2 failing

1) Luna Title component
     renders correctly in preview mode:

    AssertionError: expected '<h6.hl-title__heading.hl-title__heading--size-5>' to have CSS property 'font-size' with the value '20px', but the value was '25.008px'
    + expected - actual

    -'25.008px'
    +'20px'
    
    ...

2) Luna Title component
     renders correctly in edit mode:

    Timed out retrying after 4000ms
    + expected - actual

    { 'sling:resourceType': 'luna/components/lunatitle',
       title: 'New heading',
       showSubtitle: 'true',
    -  overlineSize: 'hl-title__heading--size-5',
       subtitle: 'New overline text',
       'jcr:primaryType': 'nt:unstructured',
       headingLevel: 'h1',
       headingSize: 'hl-title__heading--size-2' }
```

> Firstly, you updated the default font size of the overline to ensure consistency with the design. The first test detected the change and failed as a result. Secondly, you added a new property to the dialog for the component. It was recognised as well.

Reviewer: ```suggestion
Firstly, you updated the default font size of the overline to ensure consistency with the design. The first test detected the change and failed as a result. Secondly, you added a new property to the dialog for the component. It was recognized as well.
```

> ### Update functional tests

Reviewer 2: I'm missing here a couple of reasons why we do have those tests. My impression now is:
- there are tests and they failed
- you must update them, because they fail.

How about pointing a couple of reasons why we do have those tests:
- we introduced tests to automatically catch any regression
- thanks to the tests we have more confidence when introducing changes
- we believe that this approach enables us to spend less time on manual testing and regression fixes and more time on developing new features

Reviewer 3: @malaskowski thanks. I update it.

> As functional tests fail due to the changes, you should adjust related assertions. They are placed in file `tests/end-to-end/tests/lunatitle.cy.ts`. 

Reviewer 2: When functional test fail you should check why :) If this is expected, then you should adjust the tests.
Even better (but that's not the case for the quick-start guide for CMS developers) is to start with changing the test, so it fails and then adjusting the implementation so that the test passes (and finish with the code refactoring when it is covered with tests).

The first test checks the font size for the overline text. There are two component intances validated. Thus, you need to update assertions for both of them as follows.

```typescript
    cy.getByTestId('component_title1')
      .findByTestId('overline')
      .should('have.css', "font-size", "25.008px")
      .should('have.text', 'Additional overline text filled')
```

```typescript
    cy.getByTestId('component_title2')
      .findByTestId('overline')
      .should('have.css', "font-size", "25.008px")
      .should('have.text', 'Resized to 6 cols on L breakpoint')
```

> The second test validates the dialog for the component. Update the test to recognise the new input field.

Reviewer: ```suggestion
The second test validates the dialog for the component. Update the test to recognize the new input field.
```

```typescript
    cy.request(
      '/content/luna-test/pages/LunaTitle/jcr:content/rootcontainer/maincontainer/pagesection/title.json'
    )
      .its('body')
      .should('deep.eq', {
        'sling:resourceType': 'luna/components/lunatitle',
        title: 'New heading',
        showSubtitle: 'true',
        overlineSize: 'hl-title__heading--size-5',
        subtitle: 'New overline text',
        'jcr:primaryType': 'nt:unstructured',
        headingLevel: 'h1',
        headingSize: 'hl-title__heading--size-2'
      });
```

> ### Running functional tests again

Reviewer: ```suggestion
### Run functional tests again
```

Now, you can execute the functional tests again.

```shell
mvn -f tests/content/pom.xml clean install -P autoInstallPackage
npm run-script test --prefix tests/end-to-end
```

Both tests should passs this time. You should get a report like the one below.
```
Running:  lunatitle.cy.ts                                                                 (1 of 1)
 
 
Luna Title component
  ✓ renders correctly in preview mode (1030ms)
  ✓ renders correctly in edit mode (2774ms)

2 passing (4s)

```

Congratulations! You updated the component, and it passed tests.

## Part E: Use the new component

!!! info "Hint"

>     If you need help to navigate inside WebSight see the general [getting started](/docs/getting-started/quick-start/index.md) for details.

Reviewer: ```suggestion
    If you need help to navigate inside WebSight, see the general [getting started](/docs/getting-started/quick-start/index.md) for details.
```

The page owner can use the updated component now. Let's check it too.

> Run WebSight, open _Luna_ space and edit home page. Find the _Title_ having text `Meet our New Grand Luxor Jewelry Collection`. 

Reviewer 2: I'd skip writing about `Run WebSight` - it should be already running according to the tutorial.
Maybe
> Navigate to the spaces panel and open _Luna_ space...

would be better?

Reviewer 3: Previous parts focus on code changes. The goal here is to provide a context on where to execute the next steps. I would keep it.

Reviewer 3: Open replaced by Switch

Find the _Luna Title_ in the component tree on the left. Drag and drop the component on the page just below the orginal one. 

![Luna Title added](luna-title-component.png)

Edit properties of the _Luna Title_:

1. set _Heading size_ to `XL`
1. set _Heading text_ to `New Grand Luxor Jewelry Collection`
1. enable overline text
1. set _Overline size_ to `L`
1. set _Overline text_ to `Meet our` 

![Luna Title Dialog Properites](luna-title-dialog-properties.png)

> Submit changes. The title should look like expected now. You can delete the orginal _Title_ comonent to finalize the change.

Reviewer: ```suggestion
Submit changes. The title should look like expected now. You can delete the orginal _Title_ component to finalize the change.
```
